### PR TITLE
TG-2667 Refactor and expose floating point to java conversion in expr2java.h

### PR DESCRIPTION
**Pull request by LAJW**
This change exposes new `floating_point_to_java_string` function. It lets the user convert floating point numbers to Java-compatible strings without having to cast them to expression and creating an
unnecessary namespace beforehand. Also, prints Java floats in decimal format if no data is lost in the process, otherwise falls back to printing hex.

### CODE REVIEW

File: src/java_bytecode/expr2java.cpp
Comment (smowton) [inline src/java_bytecode/expr2java.cpp]: Add comment noting that this coerces NaNs to the canonical NaN

File: src/java_bytecode/expr2java.h
Comment (smowton) [inline src/java_bytecode/expr2java.h]: Does `==` work for NaNs in C? I think you might need to use `std::isnan`. This would also remove the obligation on the caller to coerce NaNs to this particular bitpattern.

File: src/java_bytecode/expr2java.h
Comment (thk123) [inline src/java_bytecode/expr2java.h]: Don't we need a `f` if float?

File: src/java_bytecode/expr2java.h
Comment (smowton) [inline src/java_bytecode/expr2java.h]: Note Peter's request that hexfloats be followed by a comment with the decimal repr

File: src/java_bytecode/expr2java.h
Comment (smowton) [inline src/java_bytecode/expr2java.h]: Aren't you missing the `f` suffix for floats?

**Review by smowton — COMMENTED** [review]
I think your rewriting has introduced some bugs, but the basic concept is ok

File: src/java_bytecode/expr2java.h
Comment (thk123) [inline src/java_bytecode/expr2java.h]: In the old code the following check was used:

```
 bool is_java_float=(src.type()==java_float_type());
```

I haven't completely understood what the above code would do, but is it not the case than on a platform where double is 32bits wide, `ieee_repr.is_double()` will return true for when the input has a width of 32bits (i.e. a Java float)

File: src/java_bytecode/expr2java.h
Comment (thk123) [inline src/java_bytecode/expr2java.h]: :+1: though is the stod -> double the same as Javas compilers?

**Review by thk123 — CHANGES_REQUESTED** [review]
This of needs some tests specifically around examples where the output should fall back to the hex output.

File: src/java_bytecode/expr2java.h
Comment (LAJW) [reply src/java_bytecode/expr2java.h]: Java reads it with the exception of Infinity/NaN

File: src/java_bytecode/expr2java.h
Comment (tautschnig) [inline src/java_bytecode/expr2java.h]: As for NaN above: use `std::isinf`

File: src/java_bytecode/expr2java.h
Comment (tautschnig) [inline src/java_bytecode/expr2java.h]: What about `std::to_string` instead of going via a stream?

File: src/java_bytecode/expr2java.h
Comment (martin-cs) [reply src/java_bytecode/expr2java.h]: This is a subtle issue but is important to get right because "x + 0.0" and "x + 0.0f" can be different!

File: src/java_bytecode/expr2java.h
Comment (martin-cs) [reply src/java_bytecode/expr2java.h]: IIRC the Java standard is quite restrictive about the link between float and double and the IEEE-754 binary32 and binary64.

File: src/java_bytecode/expr2java.h
Comment (martin-cs) [reply src/java_bytecode/expr2java.h]: Yea... I originally recommended that we used hexfloats to avoid having to deal with the vagaries of string to decimal and differences between different platform's implementations.  I mean -- do what you want but from my experience of debugging this kind of thing, keep the data as hex format and add a comment for the decimal string.

File: src/java_bytecode/expr2java.h
Comment (smowton) [inline src/java_bytecode/expr2java.h]: Noting this is still incorrect

Comment (smowton) [reply src/java_bytecode/expr2java.h]: Also style nitpick: no spaces between `if` and `(`

File: src/java_bytecode/expr2java.h
Comment (smowton) [inline src/java_bytecode/expr2java.h]: Noting still needs comment with decimal repr when using hexfloat

**Review by smowton — COMMENTED** [review]
Noting outstanding comments that show as outdated due to other fixes applied in the meantime

**Comment by LAJW** [comment]
I've addressed all changes, appended comments to the hex representation, used `std::isnan` and `std::isinf`, requesting re-review.

File: src/java_bytecode/expr2java.h
Comment (smowton) [inline src/java_bytecode/expr2java.h]: `std::stod(decimal) == value;`

File: src/java_bytecode/expr2java.h
Comment (smowton) [inline src/java_bytecode/expr2java.h]: This is backwards I think, we should use hexfloat if the decimal representation is **lossy**

File: src/java_bytecode/expr2java.h
Comment (smowton) [inline src/java_bytecode/expr2java.h]: `" */"`

**Review by smowton — CHANGES_REQUESTED** [review]

**Comment by smowton** [comment]
Also wants a quick unit test per @thk123, which would expose these bugs ;)

**Comment by LAJW** [comment]
@smowton @thk123 I've added unit tests and pushed the version with all bugs fixed. Also, I've created a [PR on test-gen](https://github.com/diffblue/test-gen/pull/1484).

File: src/java_bytecode/expr2java.h
Comment (tautschnig) [inline src/java_bytecode/expr2java.h]: Why are these assertions relevant?

Comment (LAJW) [reply src/java_bytecode/expr2java.h]: Because people were complaining that in some fictitious system float is not 32 bit and double is not 64 bit.

Comment (tautschnig) [reply src/java_bytecode/expr2java.h]: I don't know of any such system, but let's just assume it existed: why would it matter? That is, which piece of code relies on `float` being a 32-bit type.

Comment (smowton) [reply src/java_bytecode/expr2java.h]: I guess correspondence between C++ float/double and Java float/double. Annoyingly, C99 nails them down to match IEEE 754, but C++11 does not.

File: unit/java_bytecode/expr2java.cpp
Comment (smowton) [inline unit/java_bytecode/expr2java.cpp]: Should also check the negative zeroes.

**Review by smowton — APPROVED** [review]
One missing corner case to test, otherwise lgtm.

File: src/java_bytecode/expr2java.h
Comment (tautschnig) [reply src/java_bytecode/expr2java.h]: Why do we need to have a correspondence between Java and C++ float/double? We are generating strings, after all?

Comment (thk123) [reply src/java_bytecode/expr2java.h]: My reading of the above code (which may be wrong) was that the `float_type` is being determined by the compiler based on the return of `ieee_floatt::to_float` which returns a `float`. If float was 64 bit, then is_float would be true, but the value would actually correspond to a Java double.

**Review by thk123 — APPROVED** [review]
I agree with @smowton suggestion of a -0.0 test but otherwise LGTM

File: src/java_bytecode/expr2java.h
Comment (tautschnig) [reply src/java_bytecode/expr2java.h]: `float_type` will be `float` if, and only if, `ieee_repr.is_double()` returns `false`. That's completely independent of the machine that this code is being compiled on.

Comment (thk123) [reply src/java_bytecode/expr2java.h]: OK I'm convinced.

File: unit/java_bytecode/expr2java.cpp
Comment (LAJW) [reply unit/java_bytecode/expr2java.cpp]: I've added negative zero tests.

**Comment by LAJW** [comment]
I've also fixed a bug where extreme float values were throwing out-of-range errors. They were happening in casting in decimal-to-raw comparison. Hopefully, this PR is still acceptable.

File: src/java_bytecode/expr2java.h
Comment (tautschnig) [inline src/java_bytecode/expr2java.h:79]: Just curious: what is the reason to use the lambda here? The following would seem more concise to me (just 5 lines instead of 8), but I could easily be wrong:
````c++
std::ostringstream raw_stream;
raw_stream << value;
if(raw_stream.str().find('.') == std::string::npos)
  raw_stream << ".0";
const std::string decimal = raw_stream.str();
````

File: src/java_bytecode/expr2java.h
Comment (thk123) [inline src/java_bytecode/expr2java.h:92]: I don't think this is a very clear variable name, perhaps `lossless_representation`

File: src/java_bytecode/expr2java.h
Comment (thk123) [inline src/java_bytecode/expr2java.h:79]: I think a new line after each lambda call would make this function a lot more readable

Comment (LAJW) [reply src/java_bytecode/expr2java.h:79]: It is less code, but the lambda makes the mutable `raw_stream` inaccessible.

File: src/java_bytecode/expr2java.h
Comment (thk123) [inline src/java_bytecode/expr2java.h]: @tautschnig suggestion of `std::to_string(value)` might be nicer here (2 fewer lines)

File: src/java_bytecode/expr2java.h
Comment (LAJW) [reply src/java_bytecode/expr2java.h:79]: I like to group values like that with lambdas. When reading code, we're usually asking ourselves the question of "where did this value come from". Grouping values in lambdas like this makes it obvious - the results of each "code block" is plainly visible and indented. Reduces the mental load, and should the lambda become too large - it can be easily detached into a discrete function.

File: unit/java_bytecode/expr2java.cpp
Comment (thk123) [inline unit/java_bytecode/expr2java.cpp]: According to the unit test guidelines, I think this should actually be tagged with: `"[core][java_bytecode][expr2java][floating_point_to_java_string]"` 

I'd probably group then all in a 
```
SCENARIO("floating_point_to_java_string", 
  "[core][java_bytecode][expr2java][floating_point_to_java_string]")
```
And then replace each test case with a WHEN("Converting 0.0 double to string")` etc.

**Review by thk123 — COMMENTED** [review]
Some suggestions for readability

File: src/java_bytecode/expr2java.h
Comment (tautschnig) [reply src/java_bytecode/expr2java.h:79]: I guess that's where I tend to use blank lines. We probably agree to disagree on matters of style :-)

File: src/java_bytecode/expr2java.h
Comment (tautschnig) [reply src/java_bytecode/expr2java.h]: As far as I understand, `std::to_string` will yield trailing zeros.

Comment (LAJW) [reply src/java_bytecode/expr2java.h]: `std::to_string` has a different format, there's already comment on that.

File: unit/java_bytecode/expr2java.cpp
Comment (LAJW) [reply unit/java_bytecode/expr2java.cpp]: Sure, but switching to `WHEN` is pointless, since I'm not doing any mockup/teardown before and after each call.

File: src/java_bytecode/expr2java.h
Comment (LAJW) [reply src/java_bytecode/expr2java.h:79]: I think not. These kinds of lambdas serve the same purpose as new lines. The benefit of using them instead is giving smaller scopes to variable names and enabling more `const`. They are already indented, so I don't think additional new lines would help.

This isn't new, I'm just borrowing the concept from functional-ish languages (OCaml, F#), eg:

```F#
let topLevelName = 
    let nestedName1 = someExpression
    let nestedName2 = someOtherExpression
    finalExpression
```

Comment (martin-cs) [reply src/java_bytecode/expr2java.h:79]: The rest of the code base doesn't use lambdas like this and we don't need more inconsistency.  Also if I have to choose between a slightly smaller scope on variables and simpler code which works with more debuggers and profiling tools; the choice is obvious.  Please don't PR code like this.

**Comment by tautschnig** [comment]
@LAJW The Travis failure looks genuine - the unit test is failing:
`REQUIRE( floating_point_to_java_string(-5.56268e-309) == "-0x0.3ffffc7f26777p-1022 /* -5.56268e-309 */" )`

**Comment by LAJW** [comment]
DO NOT MERGE THIS YET,

I need to check if it's breaking anything in another repository.

**Pull request merged**
